### ISSUE TITLE
RFC: Export deck-kayenta view components

### ISSUE THREAD

**Issue by maggieneterval** [spinnaker/spinnaker#4195]
Currently, deck-kayenta exposes a single module usable only by deck. However, several Spinnaker community members would like to import deck-kayenta view components into standalone canary web applications that use Spinnaker’s kayenta service as a backend, and would prefer to contribute back to the project rather than forking it.

**Goals:**
- Develop deck-kayenta as a library rather than a tightly coupled dependency of deck.
- Unblock community members who are using only Spinnaker’s canary service from importing and contributing to higher-level canary view components.

**Non-goals:**
- Create a Spinnaker component library of atomic view elements to be shared among all deck modules.

**Proposal:**
- Change deck-kayenta’s build to decouple it from deck’s and make it usable as a library, possibly replacing Webpack with Rollup. This will include correctly publishing the package’s Typescript types.
- Identify which components to expose, and refactor each such that there is an exportable presentational component separate from any data fetching and side effects. This will hopefully end up being a good code-health refactor, as separating presentational and container components confers other benefits such as greater reusability and easier testing. This can be done incrementally.
- Use semantic versioning to communicate changes. To date, deck-kayenta has never experienced a major or minor version bump because it only exposes the single module. However, exposing individual components will greatly increase the API surface area, so we will need to be careful to communicate breaking changes.

**Risks:**
- Development velocity might be impeded by having to consider non-deck consumers of components. Additionally, PRs submitted by non-deck users might need to be manually tested with deck, increasing code review burden.
- Shared components that are more complicated than atomic elements like buttons have a tendency toward “prop explosion” as use cases diverge. However, only exporting the simplest view components does not address the use case described above. Hopefully we’ll get a better sense of the right level at which components should be shared, and can break them down or use patterns like [Render Props](https://reactjs.org/docs/render-props.html) to mitigate this.

All feedback is welcome!

**Comment by bpowell** [comment]
Would this be more of a component library like https://material-ui.com/ that would be used by deck and deck-kayenta? That way there is a consistent view across components?

**Comment by maggieneterval** [comment]
@bpowell I would also love to create a Spinnaker component library, but that is out of scope for this proposal (see "Non-goals").

**Comment by bpowell** [comment]
Woops, sorry! I got a little too excited, I think having a component library would really help. But it sounds like this will be a good first step to get that going

**Comment by erikmunson** [comment]
I’m generally very positive on this direction! It’d be amazing to enable some targeted use cases outside Deck without forcing folks to break away from upstream entirely.

A few things I’ve been noodling on in no particular order:
* I think it’s important to clarify what parts of the ‘deck-kayenta as a library’ story make sense for the project. In my view there’s two facets: one is allowing people to /build and import/ deck-kayenta pieces without involving the core deck codebase. Another is making the codebase and overall UI capable of operating independently from deck. I think we’re mostly aligned that the build + import part is something we want to do well — the other facet feels more nuanced and we’ll have to strike a balance
* I like the semantic versioning part, and I’ll be the first to admit we don’t do a great job of that today across our repos. That said, I think it’s worth being humble about our capability to identify breaking changes for folks using components totally outside of our world view. It’s likely (if not certain) we’d still break non-deck users in minor / patch releases, despite our best efforts to predict breakage. To make usage outside of the deck ecosystem work, we’ll need to acknowledge that the final responsibility of verifying changes will have to rest with the teams using deck-kayenta outside deck, and the rate of breakage will be higher than the equivalent rate for deck users
* I’d propose there’s three dimensions we should evaluate when deciding whether it makes sense to enable standalone usage of a given component / view / area:
	* **Value over custom implementation**: how hard / practical is it to write some custom code and keep it up to date as upstream moves forward, and how does that compare to the cost of implementing non-deck support for that thing in the upstream codebase?
	* **Expected rate of upstream change**: how often are we going to improve and change a given area in upstream? The more frequently we’ll be making upstream improvements, the more real the risk of losing development velocity due to lots of indirection/generalization for standalone use cases
	* **Ability to reason about generalized code**: at its extreme, trying to make deck-kayenta totally agnostic to deck would likely result in an unmaintainable level of generalization. We should strive to find a sweet spot where users outside deck get real and useful value while avoiding difficult to reason about code and patterns. For example: how do we handle styling components outside of deck where the look and feel should be different? Some answers imply more indirection than others, with varying levels of developer cost

At the end of the day deck-kayenta is inherently tied to deck because it’s a UI codebase that must share a similar look, feel, and set of interaction patterns + concepts — a very different proposition from the Kayenta API. Knowing that, I believe if we’re self-aware, set clear expectations, and think deeply about the tradeoffs we’ll end up doing great work together and enabling tons of innovation inside and outside deck 🚀

**Comment by maggieneterval** [comment]
Appreciate the incredibly thoughtful and perfectly worded notes, @erikmunson! +1 to all points.

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by maggieneterval** [comment]
Closing this for the time-being, as we've learned that Deck's build would need to be heavily rewritten to make this possible

**Comment by anotherchrisberry** [comment]
Hi @maggieneterval, I know this was closed over a year ago, but do you remember why Deck's build would need to be reworked for this? Or who was looking at it?

**Comment by maggieneterval** [comment]
@anotherchrisberry I think @fieldju had tried to put together a POC a while back, but to be honest I don't remember the specifics of what was a blocker.
